### PR TITLE
commons #8 extractFieldValue() should extract from Scala specific fields

### LINKED ISSUE #8: extractFieldValue should accept scala field names (not java ones)
Currently, `extractFieldValue` is using Java reflection to extract the fields. Unfortunately depending on Scala version some Scala fields have different names when compiled and therefore this function will work differently in different Scala versions.

More specifically Scala 2.12+ seems to be using short names (the Scala field name is equal to Java field name) and the older version seems to be using longer names.

Check [Scala 2.12 release notes](https://www.scala-lang.org/news/2.12.0/#scala-212-overview) for additional info.

### PR DESCRIPTION
**Pull request by wajda**
fixes #8 

This PR addresses Java/Scala reflection interoperability issue that is caused by expanded field names generated by certain Scala constructs, lambdas etc.

See unit tests

### CODE REVIEW

File: src/main/scala/za/co/absa/commons/reflect/ReflectionUtils.scala [hunk 1]
```diff
@@ -103,36 +101,81 @@ object ReflectionUtils {
...
+        }
+      }
+      .map(fld => {
+        fld.setAccessible(true)
+        fld.get(o).asInstanceOf[B]
+      })
+  }
+
+  private def expandedFieldName(c: Class[_], fieldName: String): String = {
+    s"${c.getName.replace('.', '$')}$$$$$fieldName"
```
Comment (cerveada) [inline src/main/scala/za/co/absa/commons/reflect/ReflectionUtils.scala:163]:
```suggestion
    c.getName.replace('.', '$') + "$$" + fieldName
```

**Review by cerveada — DISMISSED** [review]

**Review by cerveada — APPROVED** [review]

**Review by yruslan — DISMISSED** [review]

File: src/test/scala/za/co/absa/commons/reflect/ReflectionUtilsSpec.scala [hunk 2]
```diff
@@ -37,37 +39,72 @@ class ReflectionUtilsSpec extends AnyFlatSpec with Matchers {
...
 
-  "extractFieldValue()" should "extract from a field using a provided class tag" in {
+  it should "extract from a field using a provided class tag" in {
     val subBar = new Bar("Pi", 3.14) {}
     ReflectionUtils.extractFieldValue[Bar, String](subBar, "a") shouldEqual "Pi"
   }
 
-  "ModuleClassSymbolExtractor" should "extract objects" in {
+  it should "extract from a private field used in a lambda" in {
+    class A(x: Int) { () => x }
```
Comment (yruslan) [inline src/test/scala/za/co/absa/commons/reflect/ReflectionUtilsSpec.scala:81]: Just out of a curiosity. How does this lambda affect fetching the field?

Comment (wajda) [reply src/test/scala/za/co/absa/commons/reflect/ReflectionUtilsSpec.scala:81]: It's due to the lambda encoding. For that code the Scala compiler generates a "hidden" field to store the constructor parameter `x` value. The field has a so called _expanded_ name and looks like `za.co.absa.commons.reflect.ReflectionUtilsSpec$$anonfun$8$A$2.x`

I won't pretend I fully comprehend the reasoning behind that, especially because adding `def y = x` to that class above would result in the field name being just `x` instead of that beast, but that's what we've got.

**Comment by wajda** [comment]
Withdrawing the PR due to AbsaOSS/spline-spark-agent#138

**Review by yruslan — APPROVED** [review]

File: src/main/scala/za/co/absa/commons/reflect/ReflectionUtils.scala [hunk 3]
```diff
@@ -94,47 +93,123 @@ object ReflectionUtils {
...
-      try c.getDeclaredField(fieldName)
-      catch {
-        case e: NoSuchFieldException =>
+  def extractFieldValue[A: ClassTag, B](o: AnyRef, fieldName: String): B = {
+    @tailrec
+    def reflectClassHierarchy(c: Class[_]): Option[_] =
+      if (c == classOf[AnyRef]) None
+      else {
+        val maybeValue: Option[Any] = reflectClass(c)
+        if (maybeValue.isDefined) maybeValue
```
Comment (dk1844) [inline src/main/scala/za/co/absa/commons/reflect/ReflectionUtils.scala:114]: wouln't `getOrElse` be more readable?

**Review by dk1844 — DISMISSED** [review]
Wow

**Review by cerveada — DISMISSED** [review]

File: src/main/scala/za/co/absa/commons/reflect/ReflectionUtils.scala (hunk 3 above)
Comment (wajda) [reply src/main/scala/za/co/absa/commons/reflect/ReflectionUtils.scala:114]: It would, but I wanted it to be tail recursive

File: src/test/scala/za/co/absa/commons/reflect/ReflectionUtilsSpec.scala [hunk 4]
```diff
@@ -16,18 +16,21 @@
...
 import scala.reflect.runtime.universe._
 
-class ReflectionUtilsSpec extends AnyFlatSpec with Matchers {
+class ReflectionUtilsSpec extends AnyFlatSpec with Matchers with MockitoSugar {
 
   behavior of "ReflectionUtils"
+  behavior of "compile()"
 
-  "compile()" should "compile a given Scala code block and return an eval() function" in {
+  it should "compile a given Scala code block and return an eval() function" in {
```
[the remark this answers is not in the document]
Comment (wajda) [reply src/test/scala/za/co/absa/commons/reflect/ReflectionUtilsSpec.scala:33]: Merely string literal de-duplication. No need to copy/paste or retype the method name when adding more tests, less room for a typo. And it also kinda outlines the test-case groups boundaries in a flat spec structure.
E.g. 
```scala
behavior of "myFunction()"
it should "do X"
it should "do Y"

behavior of "anotherFunction()"
it should "do Z"
```

File: src/main/scala/za/co/absa/commons/reflect/ReflectionUtils.scala [hunk 5]
```diff
@@ -94,47 +93,123 @@ object ReflectionUtils {
...
+    @tailrec
+    def collect(ifs: Set[Class[_]], cs: Iterable[Class[_]]): Set[Class[_]] =
+      if (cs.isEmpty) ifs
+      else {
+        val c0 = cs.head
+        val cN = cs.tail
+        val ifsUpd = if (c0.isInterface) ifs + c0 else ifs
+        val csUpd = cN.toSet ++ (c0.getInterfaces filterNot ifsUpd) ++ Option(c0.getSuperclass)
+        collect(ifsUpd, csUpd)
+      }
```
[the remark this answers is not in the document]
Comment (wajda) [reply src/main/scala/za/co/absa/commons/reflect/ReflectionUtils.scala:199]: Thanks. It was a leaf over from the original pattern matching (`case head :: tail`) that doesn't work for sets.

**Review by cerveada — APPROVED** [review]

**Review by yruslan — APPROVED** [review]
